labs/b10.md

```diff
@@ -0,0 +1,157 @@
+as we will be using the `ssh info` command to check you off for the lab, and the command only works if the `decal-checkoff` user has been
+added properly.
+
+## Backups
+
+We'll be briefly practicing using `rsync`, a powerful file transfer tool often used in backup scenarios. It is difficult to 
+overstate the degree to which `rsync` and the rsync delta-transfer algorithm are used in the real world.
+
+On a superficial level, `rsync` works similarly to [`scp`](https://linux.die.net/man/1/scp) - it copies files from `src` to `dst`, optionally over the network,
+over SSH. What makes `rsync` in particular so powerful is its ability to keep two directories synchronized
+while transferring little beyond the absolute difference in files between the two directories. It does this by calculating 
+the differences between blocks in a file, and only transferring the difference (the 'delta'). By default, it assumes that a file is unchanged if its last modified time is not newer.
+
+The basic syntax of `rsync` is `rsync [options] [[user]@host:]source [[user]@host:]dest` - you can either transfer files
+in the local filesystem, between the local system and a network host, or between network hosts. 
+
+`rsync` has dozens of options, but the most common are the following:
+  * `-r` `--recursive`: recursively transfer all files and subdirectories in the source
+  * `-v` `--verbose`: verbosely print out file names and transfer speeds
+  * `-P` `--partial --progress`: keep partially transferred files if the transfer is interrupted, show progress of file transfers
+  * `-a` `--archive`: combination option, implies `-r`, and preservation of most file metadata properties
+  * `-u` `--append`: append data to the end of files on the destination that are smaller than the corresponding
+                     file on the source (useful for log files, dangerous if used on files that do not only grow by appending data)
+  * `-z` `--compress`: compress files before transfer (Only useful for `rsync`ing over the network)
+                     
+Suppose you keep all your homework in a directory "school", in your home directory:
+```
+user@machine [~/school] $ ls
+cs198-8
+cs61a
+cs262
+some_r1a
+ds8
+```
+and you want to back everything to your student VM. A command like:
+
+    $ rsync -rvPaz ~/school user@user.decal.xcf.sh:school-backup/
+    
+would result in a copy of your `school` directory being made inside the `school-backup` directory on your student VM.
+
+Play around with `rsync` yourself. We've provided some test files [here](https://decal.ocf.berkeley.edu/static/b10/b10.tar.gz). (hint: use `wget`).
+After extracting the archive, (`tar xzvf b10.tar.gz`), you'll see two directories with over 100MB of files in them.
+
+* What happens if you try `rsync`ing `dir1` to `dir2`? (hint: it should finish almost instantaneously). Why might this be the case?
```

Comment:
I couldn't get this to work, I think I was doing it wrong though. I tried doing `rsync -rvPaz ./dir1 ./dir2` initially, which just copied the contents of `dir1` into `dir2/dir1`, which wasn't the intended effect but was still pretty fast. If I did `rsync -rvPaz ./dir1/* ./dir2/`, then that worked, but because the timestamps were different between files, it didn't entirely work and copied across the whole file the first time, but subsequent times didn't copy. If I ran rsync with `rsync -crvazP ./dir1/* ./dir2/` however (`-c` is `--checksum`), that worked to not copy anything on the first run, which was good, but I feel like nobody will run it with that option.